File: src/hard-145/iteration.ts
import { BinaryTreeNode } from 'data-structures/binary-tree'

export function postorderTraversal(node: BinaryTreeNode<number> | null) {
  const answer: number[] = []
  const stack: (BinaryTreeNode<number> | null)[] = []
  if (node !== null) {
    stack.push(node)
  }

  while (stack.length > 0) {
    const top = stack.pop()

    if (top !== null && top !== undefined) {
      // 按照后序遍历的相反顺序压入栈中，使得在出栈时，各个节点是按照后序遍历的顺序出栈
      stack.push(top)
      stack.push(null)
      if (top.right) {
        stack.push(top.right)
      }
      if (top.left) {
        stack.push(top.left)
      }
    } else {
      answer.push((stack.pop() as BinaryTreeNode<number>).value)
    }
  }

  return answer
}

File: src/medium-94/iteration.go
package solution

import (
	structures "leetcode-solutions/data-structures"
)

// IterativeInOrderTraversal is medium 94 solution
func IterativeInOrderTraversal(root *structures.BinaryTreeNode) []int {
	stack := []*structures.BinaryTreeNode{}
	answer := []int{}

	if root == nil {
		return answer
	}

	current := root
	for current != nil || len(stack) > 0 {
		for current != nil {
			stack = append(stack, current)
			current = current.Left
		}
		latest := len(stack) - 1
		current = stack[latest]
		stack = stack[:latest]
		// type assertion
		// https://tour.golang.org/methods/15
		answer = append(answer, current.Val.(int))
		current = current.Right
	}

	return answer
}
